### PR TITLE
WT-15148 Use conditional variable to sync threads

### PR DESCRIPTION
**Pull request by whoisong**
The changes introduce new conditional variables to the timestamp_abort test: they are needed to avoid the timeout issue faced. Without them, the checkpoint thread may run without the test being ready for it. Each checkpoint run can be slow due to the pressure on the system and lead to timeout issues. With the condition variables in place, the checkpoint thread only starts when the test is ready which greatly reduces the chances to reach the timeout limit.

### CODE REVIEW

File: test/csuite/timestamp_abort/main.c [hunk 1]
```diff
@@ -105,6 +105,13 @@ static int recover_and_verify(uint32_t backup_index, uint32_t workload_iteration
 extern int __wt_optind;
 extern char *__wt_optarg;
 
+static struct {
+    bool tmr_ready_to_go;
+    WT_CONDVAR *cond_tmr_thrd;
+    bool ckpt_ready_to_go;
+    WT_CONDVAR *cond_ckpt_thrd;
+} thrd_sync_conds = {false, NULL, false, NULL};
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: Let's keep `checkpoint` as `ckpt`, but we can keep the full words for `timer` and `thread.
```suggestion
static struct {
    bool timer_ready_to_go;
    WT_CONDVAR *timer_cond;
    bool ckpt_ready_to_go;
    WT_CONDVAR *ckpt_cond;
} thread_sync_conds = {false, NULL, false, NULL};
```

File: test/csuite/timestamp_abort/main.c [hunk 2]
```diff
@@ -340,9 +347,17 @@ thread_ts_run(void *arg)
 
     td = (THREAD_DATA *)arg;
     conn = td->conn;
-
     testutil_check(conn->open_session(conn, NULL, NULL, &session));
-
+    /* While is to wait for all written thread reached the commit. */
+    /* Using while is to avoid spurious wakeups of conditional variables */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
    /* Wait for all working threads to complete as there will be a stable timestamp available to work with. */
```

File: test/csuite/timestamp_abort/main.c [hunk 3]
```diff
@@ -340,9 +347,17 @@ thread_ts_run(void *arg)
...
+    while (!thrd_sync_conds.tmr_ready_to_go) {
+        /* Actually we don't need this value */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: It's good that you wanted to raise it, I appreciate but we can just leave it.
```suggestion
```

**Review by etienneptl — COMMENTED** [review]
Nice work! I had a first round of review, will do another one later. Left a few comments.

File: test/csuite/timestamp_abort/main.c [hunk 4]
```diff
@@ -340,9 +347,16 @@ thread_ts_run(void *arg)
...
     td = (THREAD_DATA *)arg;
     conn = td->conn;
-
     testutil_check(conn->open_session(conn, NULL, NULL, &session));
-
+    /* Wait for all working threads to complete as there will be a stable timestamp available to
+     * work with. */
+    while (!thread_sync_conds.timer_ready_to_go) {
+        bool cv_signalled;
+        /* Wait maximum 1s each round */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
        /* Under a high workload, there is no need to check very often, check every second. */
```

File: test/csuite/timestamp_abort/main.c [hunk 5]
```diff
@@ -340,9 +347,16 @@ thread_ts_run(void *arg)
 
     td = (THREAD_DATA *)arg;
     conn = td->conn;
-
     testutil_check(conn->open_session(conn, NULL, NULL, &session));
-
+    /* Wait for all working threads to complete as there will be a stable timestamp available to
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: Multiline comment format

File: test/csuite/timestamp_abort/main.c [hunk 6]
```diff
@@ -478,6 +493,15 @@ thread_ckpt_run(void *arg)
      */
     (void)unlink(ckpt_file);
     testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+    /* Wait for the stable timestamp is configured as there will make the last_checkpoint timestamp
+     * not 0 */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
    /* 
     * Wait for the stable timestamp to be configured before starting the checkpoint thread. If
     * there is no stable timestamp, we won't create the sentinel file and we will have to
     * checkpoint again which will delay the test.
     */
```

File: test/csuite/timestamp_abort/main.c [hunk 7]
```diff
@@ -478,6 +493,15 @@ thread_ckpt_run(void *arg)
...
+    while (!thread_sync_conds.ckpt_ready_to_go) {
+        bool cv_signalled;
+        /* Wait maximum 1s each round */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
        /* Under a high workload, there is no need to check very often, check every second. */
```

File: test/csuite/timestamp_abort/main.c [hunk 8]
```diff
@@ -817,8 +841,19 @@ thread_run(void *arg)
         }
 
         /* We're done with the timestamps, allow oldest and stable to move forward. */
-        if (use_ts)
+        if (use_ts) {
+            static uint32_t progress = 0;
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: Is this variable declared at the right spot? If we execute this over and over, it will be initialised again to 0?

File: test/csuite/timestamp_abort/main.c [hunk 9]
```diff
@@ -817,8 +841,19 @@ thread_run(void *arg)
...
+            if (WT_TS_NONE == active_timestamps[td->threadnum]) {
+                uint32_t current_progress = __wt_atomic_add32(&progress, 1);
+                printf("--- syncing %u/%u @ Thread: %" PRIu32 " ---\n", current_progress, nth,
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: If we end up keeping those logs, make them more explicit please.

File: test/csuite/timestamp_abort/main.c [hunk 10]
```diff
@@ -817,8 +841,19 @@ thread_run(void *arg)
...
+                  td->threadnum);
+                if (current_progress == nth) {
+                    thread_sync_conds.timer_ready_to_go = true;
+                    __wt_cond_signal((WT_SESSION_IMPL *)session, thread_sync_conds.timer_cond);
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c:861]:
```suggestion
                    /* All threads are done working and should have updated the active timestamps array with a non-zero timestamp. It's time for the timer thread to get started. */
                    thread_sync_conds.timer_ready_to_go = true;
                    __wt_cond_signal((WT_SESSION_IMPL *)session, thread_sync_conds.timer_cond);
```

File: test/csuite/timestamp_abort/main.c [hunk 11]
```diff
@@ -936,7 +972,12 @@ run_workload(uint32_t workload_iteration)
     }
 
     opts->running = true;
-
+    /* Initialize cond machines */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
    /* Initialize cond variables. */
```

File: test/csuite/timestamp_abort/main.c [hunk 12]
```diff
@@ -986,6 +1027,9 @@ run_workload(uint32_t workload_iteration)
     for (i = 0; i <= ts_id; ++i)
         testutil_check(__wt_thread_join(NULL, &thr[i]));
 
+    __wt_cond_destroy(opt_session, &thread_sync_conds.timer_cond);
+    __wt_cond_destroy(opt_session, &thread_sync_conds.ckpt_cond);
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: Why is there a comment "NOTREACHED" below? Is it actually not reached?

File: test/csuite/timestamp_abort/main.c (hunk 9 above)
Comment (mariammojid) [reply test/csuite/timestamp_abort/main.c]: I would also propose setting them to verbose logs

File: test/csuite/timestamp_abort/main.c [hunk 13]
```diff
@@ -105,6 +105,13 @@ static int recover_and_verify(uint32_t backup_index, uint32_t workload_iteration
 extern int __wt_optind;
 extern char *__wt_optarg;
 
+static struct {
+    bool timer_ready_to_go;
+    WT_CONDVAR *timer_cond;
+    bool ckpt_ready_to_go;
```
Comment (mariammojid) [inline test/csuite/timestamp_abort/main.c:112]: I see you're using `__wt_atomic_add32` to increment progress to avoid race conditions. Have you considered using something like `__wt_atomic_storebool` and `__wt_atomic_loadbool` when accessing `timer_ready_to_go` and `ckpt_ready_to_go`?

File: test/csuite/timestamp_abort/main.c [hunk 14]
```diff
@@ -817,8 +841,19 @@ thread_run(void *arg)
         }
 
         /* We're done with the timestamps, allow oldest and stable to move forward. */
-        if (use_ts)
+        if (use_ts) {
```
Comment (mariammojid) [inline test/csuite/timestamp_abort/main.c:850]: Will everything work fine if `use_ts` is set to [false](https://github.com/wiredtiger/wiredtiger/blob/1466235c68ae9b121df996b192062cd76f268ddd/test/csuite/timestamp_abort/main.c#L1467-L1468)? It is possible to get timestamp and checkpoint deadlocks here?

**Review by mariammojid — COMMENTED** [review]
I am still trying to wrap my head around these changes and consider other edge cases. Jotted down some thoughts for now :)

File: test/csuite/timestamp_abort/main.c (hunk 13 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c:112]: There's no race scenario for these two bool, as it's single-write-single-read mode. 
- For write threads though there's lots of them but finally only one can reach the signal trigger. 
- For the tmr thread, only one exist now.

File: test/csuite/timestamp_abort/main.c (hunk 8 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c]: This process will only be executed once, in this test case. 
But considering potential scenarios, I'll move it to the global struct and add an initial action before write thread create.

File: test/csuite/timestamp_abort/main.c (hunk 12 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c]: Yes, because this thread is designed to be infinite try, and the only path to exit is killed by the parent process

File: test/csuite/timestamp_abort/main.c (hunk 14 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c:850]: Extra branch is added to work with this scenario.

File: test/csuite/timestamp_abort/main.c [hunk 15]
```diff
@@ -340,9 +348,16 @@ thread_ts_run(void *arg)
 
     td = (THREAD_DATA *)arg;
     conn = td->conn;
-
     testutil_check(conn->open_session(conn, NULL, NULL, &session));
-
+    /* Wait for all working threads to complete as there will be a
+     * stable timestamp available to work with. */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]: Please fix the multi line comment format.

File: test/csuite/timestamp_abort/main.c [hunk 16]
```diff
@@ -817,8 +845,32 @@ thread_run(void *arg)
...
+                     * started. */
+                    thread_sync_conds.timer_ready_to_go = true;
+                    __wt_cond_signal((WT_SESSION_IMPL *)session, thread_sync_conds.timer_cond);
+                }
+            }
             WT_RELEASE_WRITE_WITH_BARRIER(active_timestamps[td->threadnum], active_ts);
+        } else {
+            uint32_t current_progress = __wt_atomic_add32(&thread_sync_conds.workload_progress, 1);
+            /* In use_ts == false mode, thread sync is no more required for checkpoint creation.
+               Only first thread notify the checkpoint thread to start */
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
            /* When timestamps are not in use, we don't need to sync all threads before starting the checkpoint thread. Use the first thread to notify the checkpoint thread to start. */
```

File: test/csuite/timestamp_abort/main.c (hunk 12 above)
Comment (etienneptl) [reply test/csuite/timestamp_abort/main.c]: So should the destroy calls be below?

Comment (whoisong) [reply test/csuite/timestamp_abort/main.c]: `destroy calls` should be called after the threads are terminated, as L1046 all thread join returned, so here is a safe point.

File: test/csuite/timestamp_abort/main.c [hunk 17]
```diff
@@ -817,8 +845,32 @@ thread_run(void *arg)
...
+            if (current_progress == 1) {
+                printf(
+                  "Thread %" PRIu32 " reach syncing point, Trigger checkpoint. \n", td->threadnum);
```
Comment (etienneptl) [inline test/csuite/timestamp_abort/main.c]:
```suggestion
                  "Thread %" PRIu32 " reach syncing point, trigger checkpoint.\n", td->threadnum);
```

File: test/csuite/timestamp_abort/main.c (hunk 16 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c]: Done

File: test/csuite/timestamp_abort/main.c (hunk 17 above)
Comment (whoisong) [reply test/csuite/timestamp_abort/main.c]: Done

File: test/csuite/timestamp_abort/main.c (hunk 12 above)
Comment (etienneptl) [reply test/csuite/timestamp_abort/main.c]: Please move the code in the section below.

File: test/csuite/timestamp_abort/main.c (hunk 15 above)
Comment (etienneptl) [reply test/csuite/timestamp_abort/main.c]:
```suggestion
    /*
     * Wait for all working threads to complete as there will be a stable timestamp available to
     * work with.
     */
```

File: test/csuite/timestamp_abort/main.c (hunk 13 above)
Comment (etienneptl) [reply test/csuite/timestamp_abort/main.c:112]: @mariammojid please resolve if you are ok with the reply above.

Comment (mariammojid) [reply test/csuite/timestamp_abort/main.c:112]: thanks for looking into this and confirming @whoisong

File: test/csuite/timestamp_abort/main.c (hunk 14 above)
Comment (mariammojid) [reply test/csuite/timestamp_abort/main.c:850]: awesome thanks for confirming :)

**Review by etienneptl — APPROVED** [review]
LGTM! Good job getting to the root cause.

**Review by mariammojid — APPROVED** [review]
LGTM! Great work on this investigation and solution @whoisong
